client/asset/interface.go

```diff
@@ -0,0 +1,135 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package asset
+
+import (
+	"context"
+	"time"
+
+	"github.com/decred/dcrdex/dex"
+)
+
+// Wallet is a common interface to be implemented by cryptocurrency wallet
+// software.
+type Wallet interface {
+	// Available should return the total available funds in the wallet.
+	// Note that after calling Fund, the amount returned by Available may change
+	// by more than the value funded.
+	Available() (available, locked uint64, err error)
+	// Fund selects coins for use in an order. The coins will be locked, and will
+	// not be returned in subsequent calls to Fund or calculated in calls to
+	// Available, unless they are unlocked with ReturnCoins.
+	Fund(uint64) (Coins, error)
+	// ReturnCoins unlocks coins. This would be necessary in the case of a
+	// canceled or partially filled order.
+	ReturnCoins(Coins) error
+	// Swap sends the swap transaction. The Receipts returned can be used to
+	// refund a failed transaction.
+	Swap(*SwapTx) ([]Receipt, error)
+	// Redeem sends the redemption transaction.
+	Redeem(*Redemption) error
+	// SignMessage signs the message with the private key associated with the
+	// specified Coin. A slice of pubkeys required to spend the Coin and a
+	// signature for each pubkey are returned.
+	SignMessage(Coin, dex.Bytes) (pubkeys, sigs []dex.Bytes, err error)
+	// AuditContract retrieves information about a swap contract on the
+	// blockchain. This would be used to verify the counter-party's contract
+	// during a swap.
+	AuditContract(string, uint32, dex.Bytes) (AuditInfo, error)
+	// FindRedemption should attempt to find the input that spends the specified
+	// output, and returns the secret key if it does.
+	//
+	// DRAFT NOTE: FindRedemption is necessary to deal with the case of a maker
+	// redeeming but not forwarding their redemption information. The DEX does not
+	// monitor for this case. While it will result in the counter-party being
+	// penalized, the input still needs to be found so the swap can be completed.
+	// For BTC-like blockchains, every input of every block starting at the
+	// contract block will need to be scanned until the spending input is found.
+	// This could potentially be an expensive operation if performed long after
+	// the swap is broadcast. Realistically, though, the taker should start
+	// looking for the maker's redemption beginning at swapconf confirmations
+	// (at the latest) regardless of whether the server sends the 'redemption'
+	// message or not.
+	FindRedemption(ctx context.Context, txid string, vout uint32) (dex.Bytes, error)
```

Comment:
In all wallet implementations, this method will need extra scrutiny. It should be impossible for someone to spend the swap output in such a way as to make this function fail.